### PR TITLE
Add SupportLayer for batch-episode training

### PR DESCRIPTION
**Pull request by ClementWalter**

### CODE REVIEW

File: keras_fsl/models/layers/__init__.py [hunk 1]
```diff
@@ -1,5 +1,4 @@
-from tensorflow.keras import *  # noqa
```
Comment (Wirg) [inline keras_fsl/models/layers/__init__.py:1]: Je suis curieux de savoir pourquoi ?

File: keras_fsl/models/layers/centroids_similarity.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import tensorflow as tf
+from tensorflow.keras import activations
+
+from keras_fsl.models.layers.support_layer import SupportLayer
+
+
+class CentroidsSimilarity(SupportLayer):
+    """
+    Compute the similarity (distance) between all items of the input batch and the centroid of each class found in the batch.
```
Comment (Wirg) [inline keras_fsl/models/layers/centroids_similarity.py:9]: est-ce que similarity / distance est interchangeable ? Faut peut etre rendre ça plus clair

Comment (Wirg) [reply keras_fsl/models/layers/centroids_similarity.py:9]: je me demande si 2 classes même si c'est exactement la même ... genre une subclass avec juste un pass ?

File: keras_fsl/models/layers/centroids_similarity.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        """
+
+        Args:
+            kernel: similarity / distance function (x, x') => tf.float32
+            activation: add an activation function as in other keras standard layers (e.g. softmax to compute ProtoNets output).
+        """
+        super().__init__(kernel, **kwargs)
+        self.activation = activations.get(activation)
+
+    def build_support_set(self, inputs):
```
Comment (Wirg) [inline keras_fsl/models/layers/centroids_similarity.py:22]: un peu de typing ne ferait pas de mal, je pense

File: keras_fsl/models/layers/centroids_similarity.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        if not isinstance(inputs, list) or isinstance(inputs, list) and len(inputs) < 2:
+            raise ValueError(f"{self.__class__.__name__} should be called on a list of inputs [embeddings, labels_one_hot]")
```
Comment (Wirg) [inline keras_fsl/models/layers/centroids_similarity.py:24]: est-ce que 3 élements c'est normal ? attendu ?

File: keras_fsl/models/layers/__init__.py (hunk 1 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/__init__.py:1]: je ne sais pas pourquoi je voulais un jour possiblement importer des layers de keras depuis keras_fsl. Tant qu'il n'y a pas un besoin clair un pour un model builder particulier, les layers de keras sont importés depuis keras. Ça évite les confusions et les circular deps

File: keras_fsl/models/layers/centroids_similarity.py (hunk 2 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/centroids_similarity.py:9]: similarity / distance j'ai l'impression que c'est un peu la foire dans la littérature également. On parle de distance metric learning, cosine similarity. En fait ça va dépendre plutôt du `kernel` utilisé : si même classe = 1 on dira similarity, si même classe = 0 on dira distance. Je peux renommer en `CentroidsMatrix` si tu penses que c'est mieux, pour faire le pendant à `GramMatrix`

File: keras_fsl/models/layers/centroids_similarity.py (hunk 3 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/centroids_similarity.py:22]: c'est l'inputs du `call` , tu veux que j'ajoute un `functool.wraps`  ?

File: keras_fsl/models/layers/centroids_similarity.py (hunk 4 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/centroids_similarity.py:24]: j'avoue ici on n'en attend que deux je vais mettre ça stricte

File: keras_fsl/models/layers/support_layer.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        super().__init__(**kwargs)
+        self.kernel = kernel
+
+    def build(self, input_shape):
+        if isinstance(input_shape, list):
+            embedding_shape = input_shape[0]
+        else:
+            embedding_shape = input_shape
+
+        if not isinstance(self.kernel, Layer):
```
Comment (Wirg) [inline keras_fsl/models/layers/support_layer.py:18]: si c'est un layer tu n'as rien à faire ?
Si c; est pas un layer ça peut etre quoi ?

File: keras_fsl/models/layers/support_layer.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            kernel_config["init"] = {
+                **kernel_config.get("init", {}),
+                "input_shape": embedding_shape[1:],
+            }
+            self.kernel = getattr(head_models, kernel_config["name"])(**kernel_config["init"])
+
+    def get_config(self):
+        config = super().get_config()
+        config.update({"kernel": self.kernel.to_json()})
+        return config
```
Comment (Wirg) [inline keras_fsl/models/layers/support_layer.py:31]: il y a une raison de préférer ceci à `{**super().get_config(), "kernel": self.kernel.to_json()}}` ?

File: keras_fsl/models/layers/support_layer.py (hunk 5 above)
Comment (Wirg) [reply keras_fsl/models/layers/support_layer.py:18]: ok, je  viens de comprendre, c'est dans le cas où c'est un dico de config ?

File: keras_fsl/models/layers/centroids_similarity.py (hunk 2 above)
Comment (Wirg) [reply keras_fsl/models/layers/centroids_similarity.py:9]: c'est probablement une bonne idée Matrix

File: keras_fsl/models/layers/centroids_similarity.py (hunk 3 above)
Comment (Wirg) [reply keras_fsl/models/layers/centroids_similarity.py:22]: mais ça peut être n'importe quoi ? tu as l'air d'attendre une liste de tensors

File: keras_fsl/models/layers/support_layer.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    @classmethod
+    def from_config(cls, config):
+        kernel = tf.keras.models.model_from_json(config["kernel"])
+        config["kernel"] = kernel
+        return cls(**config)
+
+    def compute_output_shape(self, input_shape):
+        if isinstance(input_shape, list):
```
Comment (Wirg) [inline keras_fsl/models/layers/support_layer.py:40]: il y a pas un moyen de normaliser l'input_shape ? j'avais vu ça pas mal de fois dans tf

File: keras_fsl/models/layers/centroids_similarity.py (hunk 3 above)
Comment (Wirg) [reply keras_fsl/models/layers/centroids_similarity.py]: en fait, je me dis. Demain il y a un chercheur qui veut implem sa propre idée de layer.
Il va probablement utiliser ce pattern avec le `build_support_set` donc cette fonction est l'interface principale de ta classe.
Du coup, j'ai l'impression que tu veux normaliser dans le call est appeler ensuite

`self.build_support_set(embeddings, one_hot)`

Comment (Wirg) [reply keras_fsl/models/layers/centroids_similarity.py]: attend, c'est juste lui qui crash s'il a pas les deux ?

File: keras_fsl/models/layers/support_layer.py (hunk 5 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/support_layer.py:18]: oui, comme c'est fait dans `SiameseNets`

File: keras_fsl/models/layers/support_layer.py (hunk 6 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/support_layer.py:36]: non

File: keras_fsl/models/layers/support_layer.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        raise NotImplementedError
+
+    @tf.function
+    def call(self, inputs, **kwargs):
+        support_tensors = self.build_support_set(inputs)
+        support_set_size = tf.shape(support_tensors)[0]
+        if isinstance(inputs, list):
+            embeddings = inputs[0]
+        else:
+            embeddings = inputs
```
Comment (Wirg) [inline keras_fsl/models/layers/support_layer.py:59]: ça fait 3x que je vois ce bloque dans la même classe, je me demande si tu veux pas faire un

```python
def _normalize_input(inputs):
    if isinstance(inputs, list):
         return inputs[0]
    return inputs
```
Ça te permettrai de rajouter des erreurs ou de gérer des nouveaux cas plus facilement

File: keras_fsl/models/layers/tests/centroids_similarity_test.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from keras_fsl.models.layers import CentroidsSimilarity
+from keras_fsl.utils.tensors import get_dummies
+
+
+class TestCentroidsSimilarity:
```
Comment (Wirg) [inline keras_fsl/models/layers/tests/centroids_similarity_test.py]: faudra renommer le test aussi

File: keras_fsl/models/layers/support_layer.py (hunk 7 above)
Comment (ClementWalter) [reply keras_fsl/models/layers/support_layer.py:43]: en effet c'est un peu pénible

File: keras_fsl/models/layers/tests/centroids_matrix_test.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from keras_fsl.models.layers import CentroidsMatrix
+from keras_fsl.utils.tensors import get_dummies
+
+
+class TestCentroidsSimilarity:
```
Comment (Wirg) [inline keras_fsl/models/layers/tests/centroids_matrix_test.py:11]: à renommer du coup

File: notebooks/siamese_nets_training.py [hunk 11]
```diff
@@ -14,7 +14,7 @@
 from tensorflow.keras.optimizers import Adam
 
 from keras_fsl.models import SiameseNets
-from keras_fsl.sequences import training, prediction
```
Comment (Wirg) [inline notebooks/siamese_nets_training.py:17]: c'est cool ces notebooks, ça donne envie de les exploiter plus
